src/main/java/uk/gov/hmcts/ccd/domain/service/processor/GlobalSearchProcessorService.java

```diff
@@ -0,0 +1,202 @@
+package uk.gov.hmcts.ccd.domain.service.processor;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.ccd.config.JacksonUtils;
+import uk.gov.hmcts.ccd.domain.model.common.CaseFieldPathUtils;
+import uk.gov.hmcts.ccd.domain.model.definition.CaseFieldDefinition;
+import uk.gov.hmcts.ccd.domain.model.definition.CaseTypeDefinition;
+import uk.gov.hmcts.ccd.domain.model.globalsearch.OtherCaseReference;
+import uk.gov.hmcts.ccd.domain.model.globalsearch.SearchCriteria;
+import uk.gov.hmcts.ccd.domain.model.globalsearch.SearchParty;
+import uk.gov.hmcts.ccd.domain.model.globalsearch.SearchPartyValue;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+@Service
+public class GlobalSearchProcessorService {
+
+    private static final String SEARCH_CRITERIA = "SearchCriteria";
+
+    private Map<String, String> namesToValuesMap = new LinkedHashMap<>();
+
+    public Map<String, JsonNode> populateGlobalSearchData(CaseTypeDefinition caseTypeDefinition,
+                                                          Map<String, JsonNode> data) {
+        List<uk.gov.hmcts.ccd.domain.model.definition.SearchCriteria> searchCriterias =
+            caseTypeDefinition.getSearchCriterias();
+
+        List<uk.gov.hmcts.ccd.domain.model.definition.SearchParty> searchParties =
+            caseTypeDefinition.getSearchParties();
+
+        Optional<CaseFieldDefinition> searchCriteriaCaseField = caseTypeDefinition.getCaseField(SEARCH_CRITERIA);
+
+        Map<String, JsonNode> clonedData = null;
+        if (data != null) {
+            clonedData = new HashMap<>(data);
+        }
+
+        if (searchCriteriaCaseField.isPresent()) {
+
+            SearchCriteria searchCriteria = populateSearchCriteria(clonedData, searchCriterias);
+            List<SearchParty> searchPartyList = populateSearchParties(clonedData, searchParties);
+
+            if (!searchPartyList.isEmpty()) {
+                searchCriteria.setSearchParties(searchPartyList);
+            }
+            if (!searchCriteria.isEmpty() && clonedData != null) {
+                clonedData.put(SEARCH_CRITERIA, JacksonUtils.convertValueJsonNode(searchCriteria));
+            }
+        }
+
+        return clonedData;
+    }
+
+    private SearchCriteria populateSearchCriteria(Map<String, JsonNode> data,
+                                       List<uk.gov.hmcts.ccd.domain.model.definition.SearchCriteria> searchCriterias) {
+        List<OtherCaseReference> otherCaseReferences = new ArrayList<>();
+        SearchCriteria returnValue = new SearchCriteria();
+
+        if (data != null) {
+            data.forEach((key, jsonNode) -> searchCriterias.forEach(currentSearchCriteria -> {
```

Comment:
it might be efficient to swap the loops other way.
outer (searchCriterias)  and then (data) 🤔

Reply:
fixed in b1524eb

Reply:
i guess this still need fix for loop swap

Reply:
Missed this comment - fixed in bb750ee